### PR TITLE
Complex index view for generated views

### PR DESCRIPTION
**Pull request by jerben**
I also improved the design of the created view.

2021-04-15-173306_840x783_scrot

### CODE REVIEW

**Comment by jerben** [comment]
Ready for review
Todo
- [ ] Make sure changlog reflects changes

File: sihl/src/sihl.mli [hunk 1]
```diff
@@ -359,6 +423,18 @@ module Web : sig
         [error] is the error message of the validation or decoding. *)
     type form = (string * string option * string option) list
 
+    (** [find_form name form] returns the [(value, error)] of a [form] input
+        element with the [name].
+
+        The [value] is the submitted value of the input element. The value is
+        set even if the submitted form failed to decode or validate. Use the
+        submitted values to populate the form that can be fixed and re-submitted
+        by the user
```
Comment (aronerben) [inline sihl/src/sihl.mli:432]: Missing dot

File: sihl/src/sihl.mli [hunk 2]
```diff
@@ -368,8 +444,31 @@ module Web : sig
...
+      (** [search ?filter ?sort ?limit ?offset ()] returns a subset of the whole
+          collection of [t]. The returned tuple consist of the resulting subset
+          and the total number of [t] stored.
+
+          [filter] is an optional keyword that is used to apply a filter. Only
+          items are shown where [t] contains [filter] in some form. The exact
+          implementation is open, but [filter] should be used as general search
+          keyword.
+
+          [sort] is describes whether the result is sorted in descending or
```
Comment (aronerben) [inline sihl/src/sihl.mli:456]: Without "is"

File: sihl/src/web_rest.ml [hunk 3]
```diff
@@ -55,19 +55,134 @@ module type SERVICE = sig
...
+         match query.sort with
+         | Some sort -> Uri.add_query_param uri ("sort", [ string_of_sort sort ])
+         | None -> uri)
+    |> Uri.to_string
+  ;;
+
+  let next_page (query : t) (total : int) : t option =
+    let limit = Option.value ~default:default_limit query.limit in
+    let offset = Option.value ~default:0 query.offset in
+    if limit + offset < total
```
Comment (aronerben) [inline sihl/src/web_rest.ml:149]: Should it not be `<=` here?

**Review by aronerben — CHANGES_REQUESTED** [review]
Please check

File: sihl/src/gen_repo.ml [hunk 4]
```diff
@@ -47,19 +47,30 @@ WHERE uuid = ?::uuid
         |sql}
 ;;
 
-let find_all_request =
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/gen_repo.ml:50]: Fixed.

File: sihl/src/gen_repo.ml [hunk 5]
```diff
@@ -316,6 +364,18 @@ let parameters (schema : Gen_core.schema) =
   schema |> List.map (fun _ -> "?") |> String.concat ",\n  "
 ;;
 
+let filter_fragment (schema : Gen_core.schema) =
+  let open Gen_core in
+  schema
+  |> List.filter (fun (_, type_) ->
+         match type_ with
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/gen_repo.ml:371]: I didn't know you could do that, nice!

File: sihl/src/gen_view.ml [hunk 6]
```diff
@@ -95,25 +103,150 @@ let%html table_row csrf ({{name}} : {{module}}.t) =
...
+  {|<div><h3>{{module}}s</h3><table><tbody>\|\}
     (List.cons table_header items)
     {|</tbody></table></div>\|\}
 ;;
 
+let%html search_box (query : Sihl.Web.Rest.query) =
+  {|
+<form action="/{{name}}s" method="Get">
+  <input type="text" name="filter" value="\|\}
+    (query |> Sihl.Web.Rest.query_filter |> Option.value ~default:"")
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/gen_view.ml:121]: Fixed.

File: sihl/src/gen_view.ml [hunk 7]
```diff
@@ -95,25 +103,150 @@ let%html table_row csrf ({{name}} : {{module}}.t) =
...
+;;
+
+let navigate_page label (query : Sihl.Web.Rest.query option) =
+  match query with
+  | Some query ->
+    [ [%html
+        {|
+<form style="display: inline;" action="/{{name}}s" method="Get">
+  <input type="hidden" name="filter" value="\|\}
+          (query |> Sihl.Web.Rest.query_filter |> Option.value ~default:"")
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/gen_view.ml:144]: Fixed.

File: sihl/src/gen_view.ml [hunk 8]
```diff
@@ -95,25 +103,150 @@ let%html table_row csrf ({{name}} : {{module}}.t) =
...
+</form>
+\|\}
+;;
+
+let navigate_page label (query : Sihl.Web.Rest.query option) =
+  match query with
+  | Some query ->
+    [ [%html
+        {|
+<form style="display: inline;" action="/{{name}}s" method="Get">
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/gen_view.ml:142]: They are indeed the same at the moment but the user probably wants to style them differently.

File: sihl/src/sihl.mli (hunk 1 above)
Comment (jerben) [reply sihl/src/sihl.mli:432]: Fixed.

File: sihl/src/sihl.mli (hunk 2 above)
Comment (jerben) [reply sihl/src/sihl.mli:456]: Fixed.

File: sihl/src/web_rest.ml (hunk 3 above)
Comment (jerben) [reply sihl/src/web_rest.ml]: You are right, fixed.

File: sihl/src/gen_repo.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl/src/gen_repo.ml]: Yea OCaml seems to create equality for these kinds of data constructors.

File: sihl/src/gen_view.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl/src/gen_view.ml:142]: Makes sense!

**Pull request merged**